### PR TITLE
Feat/tabs component composable state

### LINKED ISSUE #910: Tabs Phase 2: Single List
Update Tabs to support the following features:

## Overflow
If there isn't enough space in the container, additional tabs should overflow into a menu:
Tabs overflowing into a menu

##  Controllable State
Add ability for developers to observe and/or control state.

### PR DESCRIPTION
**Pull request by NicholasBoll**
Fixes #910 

## Summary

Update the Tabs component to be an official Compound Component using models, behavior hooks, and compound component utility functions.

## Breaking Changes
- `onTabsChange` is now `onActivateTab` and the signature is now:
  ```tsx
  function onActivateTab({ data: { tab: string }, state: TabsState }): void
  ```
- The `<Tabs>` component no longer accepts the `currentTab` property. The Tabs uses a model now. See story for more details

### CODE REVIEW

File: cypress/integration/Tabs.spec.ts [hunk 1]
```diff
@@ -4,13 +4,14 @@ describe('Tabs', () => {
...
-  ['Simple', 'NamedKeys'].forEach(story => {
+
+  ['Simple', 'Named Keys'].forEach(story => {
     context(`given the '${story}' story is rendered`, () => {
       beforeEach(() => {
         h.stories.load('Testing/React/Labs/Tabs', story);
       });
 
-      it('should pass accessibility checks', () => {
+      it('should pass axe checks', () => {
```
Comment (NicholasBoll) [inline cypress/integration/Tabs.spec.ts:14]: We should probably change all these to specifically call out "axe" vs "accessibility". We're checking "axe" here. The rest of the file checks accessibility.

File: cypress/integration/Tabs.spec.ts [hunk 2]
```diff
@@ -4,13 +4,14 @@ describe('Tabs', () => {
...
         cy.checkA11y();
```
Comment (NicholasBoll) [inline cypress/integration/Tabs.spec.ts:15]: I might change the name here to `cy.checkAxe`. Axe does not check all accessibility issues, mostly low-hanging fruit or obvious issues.

File: modules/_labs/tabs/react/index.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
-import Tabs from './lib/Tabs';
-import TabList from './lib/TabList';
-import Tab from './lib/Tab';
-import TabPanel from './lib/TabPanel';
+import {Tabs} from './lib/Tabs';
+import {TabList} from './lib/TabList';
+import {Tab} from './lib/Tab';
+import {TabPanel} from './lib/TabPanel';
```
Comment (NicholasBoll) [inline modules/_labs/tabs/react/index.ts:4]: We haven't talked about this one, but there is so much extra boilerplate in default exports. How are they serving us now? Components like 'Button' have to decide which button export is "default" and we have to remap default exports of almost every file that uses them. Why not always use named exports?

File: cypress/integration/Tabs.spec.ts (hunk 2 above)
Comment (willklein) [reply cypress/integration/Tabs.spec.ts:15]: I agree with this idea.

File: modules/_labs/tabs/react/lib/Tab.tsx [hunk 4]
```diff
@@ -71,59 +88,36 @@ const StyledButton = styled('button')<{isSelected: boolean}>(
...
+      {
+        id: `tab-${state.id}-${tabName}`,
+        tabIndex: !!tabName && state.currentId === tabName ? undefined : -1,
+        'aria-selected': isSelected ? true : undefined,
+        'aria-controls': `tabpanel-${state.id}-${tabName}`,
+        onClick: onSelect,
+        isSelected: isSelected,
+      },
+      elemProps
+    );
```
Comment (NicholasBoll) [inline modules/_labs/tabs/react/lib/Tab.tsx:115]: Every prop list that uses a callback needs to use `mergeProps` to make sure callbacks are handled properly. This example could have use only the `onClick` instead of all the other props... Not 100% on this one.

File: modules/_labs/tabs/react/lib/TabList.tsx [hunk 5]
```diff
@@ -1,16 +1,25 @@
...
 
-export interface TabListProps extends React.HTMLAttributes<HTMLElement> {
-  /**
-   * A list of Tab components.
+export interface TabListProps {
+  children: React.ReactNode;
+  /** Optionally pass a model directly to this component. Default is to implicitly use the same
+   * model as the container component which uses React context. Only use this for advanced use-cases
    */
-  children: Tab | Tab[];
```
Comment (NicholasBoll) [inline modules/_labs/tabs/react/lib/TabList.tsx:13]: We should not restrict children. It doesn't work very well anyways and it is perfectly valid to do the following:
```
<Tabs>
  <div> // Could add padding for the panels
    <Tabs.Panel>
```

File: modules/_labs/tabs/react/lib/TabList.tsx [hunk 6]
```diff
@@ -20,58 +29,52 @@ const TabsListContainer = styled('div')({
...
-        onFocus={resetProgrammaticFocus}
-        {...elemProps}
-      >
-        {children}
-      </TabsListInnerContainer>
-    </TabsListContainer>
+/**
+ * Reset the cursor to the active tab when the tab list looses focus
+ */
+const useResetCursorOnBlur = ({state, events}: TabsModel, elemProps = {}) => {
```
Comment (NicholasBoll) [inline modules/_labs/tabs/react/lib/TabList.tsx:40]: I'm not 100% this is a sharable hook. It depends on the `state.activeTab` which is specific to the Tabs model. If we change `activeTab` to be `activeId` or `activeCursor`, it might apply to something like a Select as well which also resets cursors when focus is lost.

File: package.json [hunk 7]
```diff
@@ -59,10 +59,10 @@
     "cssstats": "^3.3.0",
     "cypress": "^4.4.1",
     "cypress-axe": "^0.5.1",
-    "cypress-enter-plugin": "^1.0.0",
+    "cypress-keyboard-plugin": "^1.0.2",
```
Comment (NicholasBoll) [inline package.json:62]: The keyboard plugin includes more than the enter key

File: modules/_labs/select/react/lib/Select.tsx [hunk 8]
```diff
@@ -488,7 +488,8 @@ class Select extends React.Component<SelectProps, SelectState> {
         case 'Enter':
           isShortcut = true;
           if (menuVisibility === 'closed' || menuVisibility === 'closing') {
-            this.toggleMenu(true);
+            // allow the click
+            isShortcut = false;
```
Comment (NicholasBoll) [inline modules/_labs/select/react/lib/Select.tsx:492]: Fixes this error: https://dashboard.cypress.io/projects/odida5/runs/2269/test-results/b8041582-848c-4085-ae92-85e6be37458e

> `event.preventDefault()` was called on this element or a parent element and no additional functionality' simulated default actions. This prevents keyboard accessibility on <button> elements. Read more at https://github.com/cypress-io/cypress/issues/311. Use `{force: true}` to disable this check

File: cypress/integration/SelectLabs.spec.ts [hunk 9]
```diff
@@ -304,7 +304,7 @@ describe('Select', () => {
 
         context('when the space key is pressed', () => {
           beforeEach(() => {
-            cy.findByLabelText('Label').type(' ');
+            cy.findByLabelText('Label').type(' ', {force: true}); // disable event.preventDefault checks
```
Comment (NicholasBoll) [inline cypress/integration/SelectLabs.spec.ts:307]: Not 100% sure about this. The keyboard plugin does allow disabling of `defaultPrevented` checks if `force: true` is passed. This technically bypasses other checks as well like visibility or interactability, but those checks are covered in other tests. This is required because the Select uses a `<button>` element, but doesn't handle the space bar the same way a button does. A button triggers click on keyup of a spacebar, which a select triggers a click on the keydown of a spacebar key event.

File: modules/_labs/tabs/react/lib/TabList.tsx [hunk 10]
```diff
@@ -20,58 +29,50 @@ const TabsListContainer = styled('div')({
...
+    },
+    elemProps
   );
 };
 
-export default TabList;
+export const TabList = createComponent('div')({
+  displayName: 'Tabs.List',
+  Component: ({children, model, ...elemProps}: TabListProps, ref, Element) => {
+    // const [tabIndicatorRef, setDimensions] = useIndicator(tabsListRef);
```
Comment (anicholls) [inline modules/_labs/tabs/react/lib/TabList.tsx:66]:
```suggestion
    // const [tabIndicatorRef, setDimensions] = useIndicator(tabsListRef);
```

File: modules/_labs/tabs/react/lib/Tabs.tsx [hunk 11]
```diff
@@ -1,210 +1,29 @@
...
-   */
-  initialTab?: string;
-  /**
-   * Callback when a tab changes. The `name` will be the `name` prop passed into the `Tabs.Item` and
-   * `Tabs.Panel` component. If a `name` isn't provided, the value will be a string of the index of
-   * the tab.
-   */
-  onTabChange?: (name: string) => void;
+export interface TabsProps extends TabsModelConfig {
+  model?: TabsModel;
```
Comment (anicholls) [inline modules/_labs/tabs/react/lib/Tabs.tsx:13]: docs?

**Review by anicholls — COMMENTED** [review]
Minor stuff. Looking good 👍  What's the plan for adding a transition to the indicator?

**Comment by NicholasBoll** [comment]
> Minor stuff. Looking good 👍 What's the plan for adding a transition to the indicator?

@anicholls My plan is to use a transition element that is only visible during the animation. This will fix the issue where the indicator is hard to calculate based on font-rendering, RTL, container resizes, window resizes, etc.

File: modules/_labs/tabs/react/lib/cursor/hooks.tsx [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    End: 'last',
+  },
+} as const;
+
+/**
+ * Handles the roving focus behavior of a Cursor model. It does not handle the tabIndex. Use
+ * `useRovingTabIndex` for that functionality.
+ * @see https://www.w3.org/TR/wai-aria-practices/#kbd_roving_tabindex
+ */
+export const useRovingFocus = ({state, events}: CursorModel, elemProps = {}) => {
```
[the remark this answers is not in the document]
Comment (NicholasBoll) [reply modules/_labs/tabs/react/lib/cursor/hooks.tsx:30]: Behavior hooks are tied to models, so this will go where the `CursorModel` goes in the future. Our packages are currently tied to components and not to more abstract concepts (except common). Perhaps `List` and `Cursor` stay abstract concepts and common makes sense.

File: modules/common/react/lib/utils/index.ts [hunk 13]
```diff
@@ -2,8 +2,9 @@ export * from './colorUtils';
 export * from './getTranslateFromOrigin';
 export * from './makeMq';
 export * from './mergeCallback';
-export * from './mergeCallbacks';
```
[the remark this answers is not in the document]
Comment (NicholasBoll) [reply modules/common/react/lib/utils/index.ts:5]: `mergeCallbacks` was never documented and only used by Tabs. Should we still add it to the migration guide?

File: modules/_labs/tabs/react/stories/stories_testing.tsx [hunk 14]
```diff
@@ -22,24 +22,26 @@ export default {
...
-    <div css={{marginTop: spacing.m}}>
-      <Tabs.Panel>Contents of First Tab</Tabs.Panel>
-      <Tabs.Panel>Contents of Second Tab</Tabs.Panel>
-      <Tabs.Panel>Contents of Third Tab</Tabs.Panel>
-      <Tabs.Panel>Contents of Fourth Tab</Tabs.Panel>
-      <Tabs.Panel>Contents of Fifth Tab</Tabs.Panel>
-    </div>
-  </Tabs>
-);
+export const Simple = () => {
```
[the remark this answers is not in the document]
Comment (NicholasBoll) [reply modules/_labs/tabs/react/stories/stories_testing.tsx:25]: I don't think our goal is dry test stories, but ensuring components work according to specifications.

We haven't solidified what we'll do with stories yet. We want to have examples that work well with component documentation and stories that are run with tests. Ideally I'd like to test examples to ensure our examples always work, but it blurs the line between where a story goes. Do we try to put all stories in the normal doc location first and only put stories that have specific edge cases that don't match our documentation or would be confusing in a test file instead?

File: modules/_labs/tabs/react/index.ts [hunk 15]
```diff
@@ -1,8 +1,10 @@
...
 
-export {Tabs, Tab, TabList, TabPanel};
+export {Tabs};
 export default Tabs;
 export * from './lib/Tabs';
+
+export * from './lib/TabList';
+export * from './lib/Tab';
+export * from './lib/TabPanel';
+export * from './lib/useTabsModel';
```
Comment (alanbsmith) [inline modules/_labs/tabs/react/index.ts:10]: TODO: Update imports

File: modules/_labs/tabs/react/stories/stories_testing.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (anicholls) [reply modules/_labs/tabs/react/stories/stories_testing.tsx:25]: I guess I was just thinking about this in the context of visual testing, not using this for functional testing. I'm fine either way, just thought that it's less maintenance to re-use if they're identical

File: .storybook/stories/compound_components.stories.mdx [hunk 16]
```diff
@@ -0,0 +1,7 @@
+import {Meta} from '@storybook/addon-docs/blocks';
+import Contributing from '../../COMPOUND_COMPONENTS.md';
+import Markdown from '../../utils/storybook/Markdown.tsx';
+
+<Meta title="Welcome/Compound Components" />
+
+<Markdown content={Contributing} />
```
Comment (NicholasBoll) [inline .storybook/stories/compound_components.stories.mdx:7]: This adds the compound component docs to Storybook

File: .storybook/webpack-loader-redirect-mdx-to-github.js [hunk 17]
```diff
@@ -1,18 +1,15 @@
...
-    // We can't determine the root directory of the project since everyone might set a different base URL
-    // The most common relative linking is between modules, so we'll split on that
-    const [_, modulePath] = dir.split('modules');
-    const newPath = modulePath
-      ? '/modules' + path.resolve(`${modulePath}`, p2)
-      : path.resolve('/', p2);
+    const newPath = path.relative(basePath, path.resolve(dir, p2));
 
-    return `[${p1}](https://github.com/Workday/canvas-kit/blob/master${newPath})`;
+    return `[${p1}](https://github.com/Workday/canvas-kit/blob/master/${newPath})`;
```
Comment (NicholasBoll) [inline .storybook/webpack-loader-redirect-mdx-to-github.js:13]: I noticed that some links didn't work and that the logic could be simpler

File: modules/_labs/tabs/react/README.md [hunk 18]
```diff
@@ -4,7 +4,7 @@
   <img src="https://img.shields.io/badge/LABS-beta-orange" alt="LABS: Beta" />
 </a>  This component is work in progress and currently in pre-release.
 
-Composable tab components
+`Tabs` is a [Compound Components](../../../../COMPOUND_COMPONENTS.md).
```
Comment (anicholls) [inline modules/_labs/tabs/react/README.md:7]:
```suggestion
`Tabs` is a [Compound Component](../../../../COMPOUND_COMPONENTS.md).
```

File: modules/_labs/tabs/react/stories/Tabs.stories.mdx [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+
+```sh
+yarn add @workday/canvas-kit-labs-react-tabs
+```
+
+## Usage
+
+This component follows the
+[W3 Tabs specification](https://www.w3.org/TR/wai-aria-practices/#tabpanel). `Tabs` is a
+[Compound Components](../../../../COMPOUND_COMPONENTS.md), which means it can be composed in a
```
Comment (anicholls) [inline modules/_labs/tabs/react/stories/Tabs.stories.mdx:30]:
```suggestion
[Compound Component](../../../../COMPOUND_COMPONENTS.md), which means it can be composed in a
```

**Review by anicholls — APPROVED** [review]
I believe disabled tabs are missing the appropriate style change. Haven't looked at the spec though, maybe I'm wrong. Otherwise this looks good :shipit:

**Comment by NicholasBoll** [comment]
> I believe disabled tabs are missing the appropriate style change. Haven't looked at the spec though, maybe I'm wrong. Otherwise this looks good :shipit:

Good catch. I've updated and added a visual regression with a disabled tab.

File: modules/_labs/tabs/react/lib/cursor/useCursorModel.tsx [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+  /** Orientation of the list. This property determines what "next" and "previous" mean. This should
+   * be set to match the layout of UI components */
+  orientation: Orientation;
+  /** The id of the list item the cursor is pointing to */
+  cursorId: string;
+};
+
+export type CursorEvents = ListEvents & {
+  /** Directly sets the cursor to the list item by its identifier. Bypasses any checks  */
+  setCursorId(data: {id: string}): void;
```
Comment (alanbsmith) [inline modules/_labs/tabs/react/lib/cursor/useCursorModel.tsx:31]: We talked about this being `goTo`, but I don't think we landed on anything. Your call.

File: modules/_labs/tabs/react/package.json [hunk 21]
```diff
@@ -36,11 +36,11 @@
   ],
   "dependencies": {
     "@emotion/core": "^10.0.28",
-    "@emotion/styled": "^10.0.27",
     "@workday/canvas-kit-react-common": "^5.0.0-beta.0",
     "@workday/canvas-kit-react-core": "^5.0.0-beta.0"
   },
   "devDependencies": {
+    "@workday/canvas-kit-react-button": "^5.0.0-beta.0",
```
Comment (alanbsmith) [inline modules/_labs/tabs/react/package.json:43]: ah, so you did add it.

File: modules/_labs/tabs/react/stories/Tabs.stories.mdx [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+example has a tab panel with a focusable button in the tab panel.
+
+<ExampleCodeBlock code={AlternativeTabStop} />
+
+### Advanced Usage
+
+Below are some examples for more advanced / custom use cases. We expect most people won't need to
+read this section, but if you're needing to go beyond basic examples or are curious, feel free to
+explore this section. While these examples are not exhaustive, they provide additional insight into
+what's possible.
```
Comment (alanbsmith) [inline modules/_labs/tabs/react/stories/Tabs.stories.mdx:78]: This is a great paragraph. 😉

File: modules/common/react/lib/utils/index.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (alanbsmith) [reply modules/common/react/lib/utils/index.ts:5]: I doubt it was used. You might just put a one-liner saying it's a name change.

**Review by alanbsmith — APPROVED** [review]
I left a few comments, but this is good to go. Great work! 🥇

File: modules/_labs/tabs/react/package.json (hunk 21 above)
Comment (NicholasBoll) [reply modules/_labs/tabs/react/package.json:43]: It hurt

File: modules/_labs/tabs/react/stories/Tabs.stories.mdx (hunk 22 above)
Comment (NicholasBoll) [reply modules/_labs/tabs/react/stories/Tabs.stories.mdx:78]: Familiar too... 🤔

File: modules/_labs/tabs/react/lib/cursor/useCursorModel.tsx (hunk 20 above)
Comment (NicholasBoll) [reply modules/_labs/tabs/react/lib/cursor/useCursorModel.tsx:31]: Yes. I think `goTo` would work. It is hard to come up with a name that isn't too verbose, yet works in multiple contexts. This event will set the cursor to a specific point in the list, which is used by the Tabs model. In this context, `goTo` will set the focus cursor while `activateTab` will set the active tab.

Should `activateTab` be changed to `activate` and assume the `Tab` context is implied? It can be difficult with model composition to make too many assumptions about the context in which the events will be called.

Should we always or never explicitly specify model context?

Explicit model context:
- `registerListItem` - from `ListModel`
- `goToCursor` - from `CursorModel`
- `activateTab` - from `TabsModel`

Implicit model context:
- `registerItem` - from `ListModel`
- `goTo` - from `CursorModel`
- `activate` - from `TabsModel`

File: cypress/integration/Popup.spec.ts [hunk 23]
```diff
@@ -113,8 +113,7 @@ describe('Popup', () => {
           cy.findByLabelText('Popup 2').should('be.visible');
         });
 
-        // TODO Skip for now until we have a systematic approach to fix this issue
-        it.skip('should close Popup 1', () => {
+        it('should close Popup 1', () => {
```
Comment (NicholasBoll) [inline cypress/integration/Popup.spec.ts:116]: Apparently it is fixed.

File: modules/_labs/breadcrumbs/react/lib/Breadcrumbs/Dropdown/MenuItemLink.tsx [hunk 24]
```diff
@@ -33,7 +33,7 @@ const MenuItemLink = styled('a')({
 export const DropdownMenuItemLink = forwardRef(
   (
     {children, href, ...elemProps}: DropdownMenuItemLinkProps,
-    ref: React.RefObject<HTMLAnchorElement>
+    ref: React.Ref<HTMLAnchorElement>
```
Comment (NicholasBoll) [inline modules/_labs/breadcrumbs/react/lib/Breadcrumbs/Dropdown/MenuItemLink.tsx:36]: This was caught by `strictFunctionTypes`

File: modules/_labs/breadcrumbs/react/lib/Breadcrumbs/List/hooks.ts [hunk 25]
```diff
@@ -50,7 +50,10 @@ export const useBuildCollapsedList = <E extends HTMLElement>(
     const listItemNodes = listEl.current ? listEl.current.querySelectorAll('li') : [];
     const containerWidth = listEl.current ? listEl.current.clientWidth : 0;
     const listItems: Breadcrumb[] = [];
-    Children.forEach(children, (child: React.ReactElement, index) => {
+    Children.forEach(children, (child, index) => {
+      if (!React.isValidElement(child)) {
+        return;
+      }
```
Comment (NicholasBoll) [inline modules/_labs/breadcrumbs/react/lib/Breadcrumbs/List/hooks.ts:56]: This was caught by `strictFunctionTypes`. I think this is a valid issue

File: modules/_labs/pagination/react/lib/Pagination/types.ts [hunk 26]
```diff
@@ -9,7 +7,7 @@ export interface PaginationState {
 }
 
 export interface PaginationEvents {
-  setCurrentPage: React.Dispatch<React.SetStateAction<number>>;
+  setCurrentPage: (page: number) => void;
```
Comment (NicholasBoll) [inline modules/_labs/pagination/react/lib/Pagination/types.ts:10]: This was a valid catch by `strictFunctionTypes`

File: modules/_labs/menu/react/lib/Menu.tsx [hunk 27]
```diff
@@ -132,7 +132,10 @@ export default class Menu extends React.Component<MenuProps, MenuState> {
           ref={this.menuRef}
           {...elemProps}
         >
-          {React.Children.map(children, (menuItem: React.ReactElement<MenuItemProps>, index) => {
+          {React.Children.map(children, (menuItem, index) => {
+            if (!React.isValidElement(menuItem)) {
+              return;
+            }
```
Comment (NicholasBoll) [inline modules/_labs/menu/react/lib/Menu.tsx:138]: Valid catch by `strictFunctionTypes`

File: modules/_labs/tabs/react/lib/cursor/hooks.tsx [hunk 28]
```diff
@@ -55,15 +55,15 @@ export const useRovingFocus = ({state, events}: CursorModel, elemProps = {}) =>
...
+        (Object.keys(
+          orientationKeyMap[state.orientation]
+        ) as (keyof typeof orientationKeyMap[typeof state.orientation])[]).forEach(key => {
+          if (isRTL ? event.key === rightToLeftMap[key] : event.key === key) {
+            const event = orientationKeyMap[state.orientation][key];
+            focusRef.current = true;
+            events[event]?.();
           }
-        );
+        });
```
Comment (NicholasBoll) [inline modules/_labs/tabs/react/lib/cursor/hooks.tsx]: `Object.keys` returns `string[]` vs `(keyof T)[]` which is why we have to cast. Sometimes I make a simple utility function with the correct signature. Why couldn't they just update `Object.keys`?

File: modules/_labs/tabs/react/lib/cursor/useCursorModel.tsx [hunk 29]
```diff
@@ -114,7 +115,7 @@ export const useCursorModel = (config: CursorModelConfig = {}): CursorModel => {
   const [cursorId, setCursorId] = React.useState('');
   const initialCurrentRef = React.useRef('');
   const list = useListModel({
-    ...config,
+    ...(config as ListModelConfig),
```
Comment (NicholasBoll) [inline modules/_labs/tabs/react/lib/cursor/useCursorModel.tsx:118]: Narrowing the type here prevents `strictFunctionTypes` errors on the `...config` object having covariant errors between `CursorModelConfig` and `ListModelConfig` where `config` technically has callback and guard types that accept `CursorState` instead of `ListState`. Bivariance says this is okay since `CursorState` extends from `ListState`, but contravariance says it isn't. This casting prevents the need for bivariance hacks. A type error would occur if the casting was invalid, but I think this is a valid tradeoff.

Comment (NicholasBoll) [reply modules/_labs/tabs/react/lib/cursor/useCursorModel.tsx:118]: And if you don't fully understand covariance, contravariance, and bivariance, that's okay. I don't either.

File: modules/common/react/lib/utils/components.ts [hunk 30]
```diff
@@ -166,7 +166,7 @@ export function useForkRef<T>(ref1: React.Ref<T>, ref2: React.Ref<T>): (instance
  *   return <Element ref={elementRef} {...elemProps} />
  * }
  */
-export function useLocalRef<T>(ref: React.Ref<T>) {
+export function useLocalRef<T>(ref?: React.Ref<T>) {
```
Comment (NicholasBoll) [inline modules/common/react/lib/utils/components.ts:169]: I made these refs be optional in Typescript because they are technically optional. `strictFunctionTypes` surfaced this need.

File: modules/_labs/tabs/react/lib/cursor/useCursorModel.tsx (hunk 29 above)
Comment (NicholasBoll) [reply modules/_labs/tabs/react/lib/cursor/useCursorModel.tsx:118]: More info here: https://www.stephanboyer.com/post/132/what-are-covariance-and-contravariance

File: modules/common/react/lib/utils/models.ts [hunk 31]
```diff
@@ -5,14 +5,17 @@ export type Model<State, Events> = {
   events: Events;
 };
 
+// bivarianceHack is used for force bivariance of the function constraint. Without this, it will fail with `strictFunctionTypes`
+type IEvent = {[key: string]: {bivarianceHack(data?: object): void}['bivarianceHack']};
```
Comment (NicholasBoll) [inline modules/common/react/lib/utils/models.ts:9]: This bivariance hack is the same as what's used in React. This is a case for valid bivariance - I want to set a constraint of an `Events` object. It must be a record of functions that either accept an object or nothing. This is to make callback and guard signatures predictable through the `createEventMap` and `useEventMap` utility functions. `strictFunctionTypes` complain about this bivariance though since a real event will not have a signature of `(data: object) => void` and will look like `(data: { foo: string }) => void` which is a contravariance violation. But it is valid to set up constraints in this case, so a bivariance hack is necessary.

Here's a simplified example of what's going on: [Typescript Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYgrgOwMZQLxQN4CMCWA3AQwCccDkIAJApAawAoATA4AgfgC4oB7LAKwiTAAlJzxccDKAF8A2gCJchEmSSVqNOQF0AUAHpdoSLEQp0jZgU49+goWgB8UMRKj6oSAnADOEL+64IwEQESqTkTjhcADbMkQhQAO44wAAWUF5BOILwyMBxACrgvtraAGYmeQHuRBDMEDlIADz5UBAAHsAQCAx+DfZ0pQic+XYY2lBQNcBwRPGD2lIlSAEZ1bWdDA1oa3UNdOYsnBhQpVxcnBkkCADm0naojmMTUzMIC0JLK8BQALYgMNRgFwiCBts1Wh0uj1uHwBMB+jgEGA4MBhvdHuNJhBprMoIjkcAFp8EKtOj8wNs-gDBMCQHQMKdzlAAORYYjMqRCIA)

File: modules/popup/react/lib/Popper.tsx [hunk 32]
```diff
@@ -65,10 +65,7 @@ export interface PopperProps extends React.HTMLAttributes<HTMLDivElement> {
 }
 
 export const Popper = React.forwardRef<HTMLDivElement, PopperProps>(
-  ({portal = true, open = true, ...elemProps}: PopperProps, forwardRef) => {
-    const localRef = React.useRef<HTMLDivElement>(null);
-    const ref = (forwardRef || localRef) as React.RefObject<HTMLDivElement>;
-
```
Comment (NicholasBoll) [inline modules/popup/react/lib/Popper.tsx:71]: This was a whoops. `forwardRef` could be a function if it was used in a Class component. In this case, `forwardRef` would be type narrowed to a `React.RefObject` (subtype of `React.Ref`). This was caught as a contravariant violation on line 122 which is totally valid. It required me to fix this possible runtime error by refactoring how `usePopupStack` works with refs and accept `React.Ref` instead of only `React.RefObject`.

File: modules/popup/react/lib/Popper.tsx [hunk 33]
```diff
@@ -119,12 +116,12 @@ const OpenPopper = React.forwardRef<HTMLDivElement, PopperProps>(
       containerElement,
       ...elemProps
     },
-    forwardRef: React.RefObject<HTMLDivElement>
+    ref
   ) => {
     const firstRender = React.useRef(true);
     const popperInstance = React.useRef<PopperJS.Instance>();
     const [placement, setPlacement] = React.useState(popperPlacement);
-    const stackRef = usePopupStack(forwardRef, anchorElement as HTMLElement);
```
Comment (NicholasBoll) [inline modules/popup/react/lib/Popper.tsx:127]: Possible breaking change. Before, if `forwardRef` was defined, `stackRef` would be the same reference. Now they will always be different references since `usePopupStack` now uses forked refs. This is much more safe.

File: modules/popup/react/lib/usePopupStack.ts [hunk 34]
```diff
@@ -40,23 +41,22 @@ import {PopupStack} from '@workday/canvas-kit-popup-stack';
...
  */
 export const usePopupStack = <E extends HTMLElement>(
-  forwardRef?: React.RefObject<E>,
+  ref?: React.Ref<E>,
   target?: HTMLElement | React.RefObject<HTMLElement>
-): React.RefObject<HTMLDivElement> => {
-  const internalRef = React.useRef<HTMLDivElement>(null);
-  const ref = (forwardRef || internalRef) as React.RefObject<HTMLDivElement>;
+): React.RefObject<HTMLElement> => {
+  const {elementRef, localRef} = useLocalRef(ref);
```
Comment (NicholasBoll) [inline modules/popup/react/lib/usePopupStack.ts:47]: `useLocalRef` is a convenience wrapper around `useForkRef` that extracts types from the provided ref and returns accurate types that are more clear on the intent.

File: modules/popup/react/lib/usePopupStack.ts [hunk 35]
```diff
@@ -40,23 +41,22 @@ import {PopupStack} from '@workday/canvas-kit-popup-stack';
  * useCloseOnEscape(stackRef, closePopup)
  */
 export const usePopupStack = <E extends HTMLElement>(
-  forwardRef?: React.RefObject<E>,
+  ref?: React.Ref<E>,
```
Comment (NicholasBoll) [inline modules/popup/react/lib/usePopupStack.ts:44]: I'm not 100% on what to do here. I really hate that `div` elements are `HTMLDivElement` instead of `HTMLElement` which causes all kinds of type incompatibilities when you use any tag other than a `div`.

The `PopupStack` is in charge of creating the element and it makes a `div`. That ref is never meant to be applied to a React element and this will throw an error if you even attempt it. The `RefObject` returned by this function will be used by all behavior hooks that are meant to be used with stacks and is the `RefObject` used by a React Portal.

File: modules/common/react/lib/utils/models.ts (hunk 31 above)
Comment (mannycarrera4) [reply modules/common/react/lib/utils/models.ts:9]: i really appreciate the description here plus the playground!

File: modules/popup/react/lib/usePopupStack.ts [hunk 36]
```diff
@@ -66,12 +66,12 @@ export const usePopupStack = <E extends HTMLElement>(
...
       : undefined;
-    PopupStack.add({element: ref.current!, owner: targetEl});
-    const element = ref.current!;
+    PopupStack.add({element: localRef.current!, owner: targetEl});
+    const element = localRef.current!;
     return () => {
       PopupStack.remove(element);
     };
-  }, [ref, target, popupRef]);
+  }, [localRef, target, popupRef]);
```
Comment (NicholasBoll) [inline modules/popup/react/lib/usePopupStack.ts:74]: I don't think refs are technically needed anymore by the lint rule if they are locally defined, but it doesn't hurt other than unnecessary reference comparisons.

File: modules/_labs/breadcrumbs/react/lib/Breadcrumbs/Dropdown/MenuItemLink.tsx (hunk 24 above)
Comment (alanbsmith) [reply modules/_labs/breadcrumbs/react/lib/Breadcrumbs/Dropdown/MenuItemLink.tsx:36]: whoops. 😬

File: modules/popup/react/lib/usePopupStack.ts [hunk 37]
```diff
@@ -40,23 +41,22 @@ import {PopupStack} from '@workday/canvas-kit-popup-stack';
  * useCloseOnEscape(stackRef, closePopup)
  */
 export const usePopupStack = <E extends HTMLElement>(
```
Comment (NicholasBoll) [inline modules/popup/react/lib/usePopupStack.ts:43]: I'm not sure I need a generic here at all... This mostly prevents type issues for consumers if they use the wrong ref type...

File: modules/tooltip/react/spec/Tooltip.spec.tsx [hunk 38]
```diff
@@ -1,5 +1,3 @@
-/// <reference types="@testing-library/jest-dom/extend-expect" />
```
Comment (NicholasBoll) [inline modules/tooltip/react/spec/Tooltip.spec.tsx:1]: Automatically handled now and this manual reference causes TS issues

**Review by alanbsmith — APPROVED** [review]
🦅

File: modules/_labs/breadcrumbs/react/lib/Breadcrumbs/Dropdown/MenuItemLink.tsx (hunk 24 above)
Comment (NicholasBoll) [reply modules/_labs/breadcrumbs/react/lib/Breadcrumbs/Dropdown/MenuItemLink.tsx:36]: No runtime issue, just not the correct type. It was narrowed by the type override. Nothing depending on this `ref` required a `RefObject` or assumed it was. But it was nice `strictFunctionTypes` caught this as a potential problem.
